### PR TITLE
feat(auth): Add key previews to personal API keys

### PR DESCRIPTION
**Pull request by mishushakov**
## Problem

Fixes issue #22360

## Changes

- Changes PersonalAPIKey data model
- Adds function to mask keys
- Returns masked value in API
- Displays the masked value in the frontend Settings > Personal API keys:

Screenshot 2024-05-20 at 17 19 32

## Does this work well for both Cloud and self-hosted?

Works on self-hosted.

## How did you test this code?

Ran in codespaces on localhost.

### CODE REVIEW

**Comment by mishushakov** [comment]
The mask was inspired by OpenAI API keys page, let me know if you want this changed

**Review by benjackwhite — CHANGES_REQUESTED** [review]
I like the idea but we definitely don't want a migration and a separate field for it. The key is available - we should just read it and do the masking at read time instead. 

Also then we can just use the existing field

**Comment by Twixes** [comment]
Ah @benjackwhite This is not is clear in the code, but the `PersonalAPIKey.value` field is not used. We no longer store non-hashed values of keys, to prevent a leak of the table's contents from granting any malicious actor access to user accounts. So we do need a migration to facilitate this

File: posthog/models/personal_api_key.py [hunk 1]
```diff
@@ -35,11 +35,16 @@ def hash_key_value(value: str, mode: ModeType = "sha256", iterations: Optional[i
...
+def mask_key_value(value: str) -> str:
+    return f"{value[:3]}...{value[-4:]}"
+
+
 class PersonalAPIKey(models.Model):
     id: models.CharField = models.CharField(primary_key=True, max_length=50, default=generate_random_token)
     user = models.ForeignKey("posthog.User", on_delete=models.CASCADE, related_name="personal_api_keys")
     label: models.CharField = models.CharField(max_length=40)
     value: models.CharField = models.CharField(unique=True, max_length=50, editable=False, null=True, blank=True)
+    mask_value: models.CharField = models.CharField(unique=True, max_length=50, editable=False, null=True)
```
Comment (Twixes) [inline posthog/models/personal_api_key.py:47]: This should not have `unique=True`, as it's not _that_ unlikely that there will be two API keys ending with the same 4 characters.

File: frontend/src/scenes/settings/user/PersonalAPIKeys.tsx [hunk 2]
```diff
@@ -339,6 +339,12 @@ function PersonalAPIKeysTable(): JSX.Element {
                         )
                     },
                 },
+                {
+                    title: 'Key',
+                    dataIndex: 'mask_value',
+                    key: 'mask_value',
+                    render: (_, key) => key.mask_value,
```
Comment (Twixes) [inline frontend/src/scenes/settings/user/PersonalAPIKeys.tsx:346]: All existing keys will have a null `mask_value`, so we need to elegantly support that case too

**Review by Twixes — COMMENTED** [review]
A few suggestions, but in principle sounds good to me!
It's effectively a sacrifice of ~24 bits of entropy, out the key's 256 bits total, so we could also increase the key's random part to 280 bits. Let's do that in the `generate_random_token_personal()` function, by changing `generate_random_token()` to `generate_random_token(35)`

File: frontend/src/scenes/settings/user/PersonalAPIKeys.tsx (hunk 2 above)
Comment (mishushakov) [reply frontend/src/scenes/settings/user/PersonalAPIKeys.tsx]: I have added "Preview unavailable" when the `mask_value` is not set in database

**Review by Twixes — APPROVED** [review]
Made a few small tweaks – the PR looks good to me! Will wait for a second opinion from @benjackwhite though

**Review by benjackwhite — CHANGES_REQUESTED** [review]
Ah yeah - shouldn't have been trying to review this on the go 😅 Yeah we need the extra field indeed and overall this looks good. 

Only remaining issue will be the double migration. I think our CI requires a max of one migration per PR so needs to be combined into one

**Comment by mishushakov** [comment]
Hey @benjackwhite, I have updated the migration code 😄

**Review by benjackwhite — APPROVED** [review]
Amazing! Great addition!

**Comment by benjackwhite** [comment]
I could imagine some of the tests failing as they check the returned payload so that might be the final thing to get this over the line https://github.com/PostHog/posthog/blob/feat/cli-auth/posthog/api/test/test_personal_api_keys.py#L27

**Comment by mishushakov** [comment]
Have updated the tests as you pointed out, just in case 👍

**Pull request merged**